resolver/resolver.go

```diff
@@ -0,0 +1,129 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package resolver defines APIs for name resolution in gRPC.
+// All APIs in this package are experimental.
+package resolver
+
+var (
+	// m is a map from scheme to resolver builder.
+	m map[string]Builder
+	// defaultScheme is the default scheme to use.
+	defaultScheme string
+)
+
+func init() {
+	// TODO(bar) install dns resolver.
+	m = make(map[string]Builder)
+}
+
+// Register registers the resolver builder to the resolver map.
+// b.Scheme will be used as the scheme registered with this builder.
+func Register(b Builder) {
+	m[b.Scheme()] = b
+}
+
+// Get returns the resolver builder registered with the given scheme.
+// If no builder is register with the scheme, the default scheme will
+// be used.
+// If the default scheme is not modified, "dns" will be the default
+// scheme, and the preinstalled dns resolver will be used.
+// If the default scheme is modified, and a resolver is registered with
+// the scheme, that resolver will be returned.
+// If the default scheme is modified, and no resolver is registered with
+// the scheme, (nil, false) will be returned.
+func Get(scheme string) (b Builder, ok bool) {
+	b, ok = m[scheme]
+	if ok {
+		return
+	}
+	b, ok = m[defaultScheme]
+	return
+}
+
+// SetDefaultScheme sets the default scheme that will be used.
+// The default default scheme is "dns".
+func SetDefaultScheme(scheme string) {
+	defaultScheme = scheme
+}
+
+// AddressType indicates the address type returned by name resolution.
+type AddressType uint8
+
+const (
+	// Backend indicates the address is for a backend server.
+	Backend AddressType = iota
+	// GRPCLB indicates the address is for a grpclb load balancer.
+	GRPCLB
+)
+
+// Address represents a server the client connects to.
+// This is the EXPERIMENTAL API and may be changed or extended in the future.
+type Address struct {
+	// Addr is the server address on which a connection will be established.
+	Addr string
+	// Type is the type of this address.
+	Type AddressType
+	// ServerName is the name of this address.
+	// It's the name of the grpc load balancer, which will be used for authentication.
+	ServerName string
+	// Metadata is the information associated with Addr, which may be used
+	// to make load balancing decision.
+	Metadata interface{}
+}
+
+// BuildOption includes additional information for the builder to create
+// the resolver.
+type BuildOption struct {
+}
+
+// ClientConnection contains the callbacks for resolver to notify any updates
+// to the gRPC ClientConn.
+type ClientConnection interface {
+	// NewAddress is called by resolver to notify ClientConnection a new list
+	// of resolved addresses.
+	// The address list should be the complete list of new addresses.
+	// Two consecutive call with the same addresses will be noop. It's preferred
+	// that resolver doesn't call this consecutively with non-updated addresses.
```

Comment:
I wonder if we even need this sentence.  Like, if the resolver refreshes every 30 minutes and calls this, it's not a problem.  The idea was to avoid a busy-loop situation, right?  I'm not sure who would implement a resolver to continually call NewAddress, or what line of thinking might even lead to that.